cpp/src/aztec/stdlib/primitives/memory/dynamic_array.cpp

```diff
@@ -0,0 +1,277 @@
+            // field_pt bounds_check = (t2 + t2).madd(t1 - 1, -t1);
+
+            _context->create_new_range_constraint(bounds_check.normalize().get_witness_index(), _max_size);
+        }
+
+        bool_pt index_currently_invalid = bool_pt(witness_pt(_context, i >= native_size()));
+        {
+            // index_delta will be between 0 and length - 1 if index valid
+            // i.e. will pass check that index_delta < _max_size
+            field_pt index_delta = (_length - i - 1);
+
+            // reverse_delta will be between 0 and (_max_size - length) if *invalid*
+            // i.e. will pass check that reverse_delta < _max_size
+            field_pt reverse_delta = (-_length + i);
+
+            field_pt bounds_check = field_pt::conditional_assign(index_currently_invalid, reverse_delta, index_delta);
+
+            _context->create_new_range_constraint(bounds_check.normalize().get_witness_index(), _max_size);
+        }
+
+        field_pt old_value = _inner_table.read(i);
+        field_pt new_value =
+            field_pt::conditional_assign(index_currently_invalid && index_valid, default_value, old_value);
+        _inner_table.write(i, new_value);
+    }
+
+    _length = new_length;
+}
+
+/**
+ * @brief Read a field element from the dynamic array at an index value
+ *
+ * @tparam Composer
+ * @param index
+ * @return field_t<Composer>
+ */
+template <typename Composer> field_t<Composer> DynamicArray<Composer>::read(const field_pt& index) const
+{
+    const field_pt index_delta = (_length - index - 1);
+
+    if (index_delta.is_constant()) {
+        bool valid = (uint256_t(index_delta.get_value()) < _max_size);
+        if (!valid) {
+            _context->failure("DynamicArray::read access out of bounds");
+        }
+    } else {
+        _context->create_new_range_constraint(
+            index_delta.normalize().get_witness_index(), _max_size, "DynamicArray::read access out of bounds");
+    }
+
+    return _inner_table.read(index);
+}
+
+/**
+ * @brief Write a field element into the dynamic array at an index value
+ *
+ * @tparam Composer
+ * @param index
+ * @param value
+ */
+template <typename Composer> void DynamicArray<Composer>::write(const field_pt& index, const field_pt& value)
+{
+    const field_pt index_delta = (_length - index - 1);
+
+    if (index_delta.is_constant()) {
+        bool valid = (uint256_t(index_delta.get_value()) < _max_size);
+        if (!valid) {
+            _context->failure("DynamicArray::read access out of bounds");
+        }
+    } else {
+        _context->create_new_range_constraint(
+            index_delta.normalize().get_witness_index(), _max_size, "DynamicArray::read access out of bounds");
+    }
+
+    _inner_table.write(index, value);
+}
+
+/**
+ * @brief Push a field element onto the dynamic array
+ *
+ * @tparam Composer
+ * @param value
+ */
+template <typename Composer> void DynamicArray<Composer>::push(const field_pt& value)
```

Comment:
I think you forgot a range check for the new size

Reply:
The `RamTable::write` method should take care of this.

The `DynamicArray` class wraps a `RamTable` table object with a fixed maximum size. We apply an arithmetic constraint that checks the final 'index' value in our sorted list of RAM read/writes equals the maximum size (ultra_composer.cpp line 2486), if the index is out-of-bounds this constraint will be unsatisfied.